### PR TITLE
Preflight check improvement

### PR DESCRIPTION
**Pull request by LorenzMeier**
By @Zefz. @kd0aij Could you give this a try and see if the system still allows you to arm you, and if it behaves correctly if you e.g. reset the MAG0 ID param?

### CODE REVIEW

**Comment by kd0aij** [comment]
Checked out and built preflight-check-improvement for fmuV1
Forced reflash of PX4IO
Armed successfully
disarmed
```
nsh> param set CAL_MAG0_ID 0
+ CAL_MAG0_ID: curr: 73233 -> new: 0
nsh> sensors_task: NO CONFIG FOR MAG #0
sensors_task: config: 2 gyros, 0 mags, 1 accels
```
rearmed successfully
param save
reboot
continuous print on console (this should only print once, right?):
```
commander: INVAL: ARMING_STATE_INIT - ARMING_STATE_STANDBY_ERROR
```

**Comment by Zefz** [comment]
```
commander: INVAL: ARMING_STATE_INIT - ARMING_STATE_STANDBY_ERROR
```
This is actually a bug I think. When e sensor has failed at bootup, the commander enters a STANDBY_ERROR mode. When all sensors are calibrated and working again, it tries to go from STANDBY_ERROR to INIT which is not allowed.

**Comment by LorenzMeier** [comment]
@Zefz This looks like its not allowed to go from ARMING_STATE_INIT to ARMING_STATE_STANDBY_ERROR. So what fails is entering the arming error state, not leaving it.

**Comment by Zefz** [comment]
Does this need fixing before a merge?

**Comment by LorenzMeier** [comment]
It seems so - its in a weird state if the preflight check fails (or the error message on the console is wrong).

**Comment by LorenzMeier** [comment]
I have rebased this on master, and I'd like to get this one in ASAP.

File: src/modules/commander/PreflightCheck.cpp [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            return false;
+        }
+
+        int calibration_devid;
+        int devid = ioctl(fd, DEVIOCGDEVICEID,0);
+        param_get(param_find("CAL_MAG0_ID"), &(calibration_devid));
+        if (devid != calibration_devid){
+            warnx("magnetometer calibration is for a different device - calibrate magnetometer first (dev: %d vs cal: %d)", devid, calibration_devid);
+            mavlink_log_critical(mavlink_fd, "SENSOR FAIL: MAG CAL ID");
+            return false;
```
Comment (LorenzMeier) [inline src/modules/commander/PreflightCheck.cpp]: leaking file descriptor

File: src/modules/commander/PreflightCheck.cpp [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        }
+
+        int ret = ioctl(fd, MAGIOCSELFTEST, 0);
+        if (ret != OK) {
+            warnx("magnetometer calibration missing or bad - calibrate magnetometer first");
+            mavlink_log_critical(mavlink_fd, "SENSOR FAIL: MAG CHECK/CAL");
+            return false;
```
Comment (LorenzMeier) [inline src/modules/commander/PreflightCheck.cpp]: leaking file descriptor

File: src/modules/commander/PreflightCheck.cpp [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        int fd = open(ACCEL0_DEVICE_PATH, O_RDONLY);
+        int ret = ioctl(fd, ACCELIOCSELFTEST, 0);
+
+        int calibration_devid;
+        int devid = ioctl(fd, DEVIOCGDEVICEID,0);
+        param_get(param_find("CAL_ACC0_ID"), &(calibration_devid));
+        if (devid != calibration_devid){
+            warnx("accelerometer calibration is for a different device - calibrate accelerometer first");
+            mavlink_log_critical(mavlink_fd, "SENSOR FAIL: ACC CAL ID");
+            return false;
```
Comment (LorenzMeier) [inline src/modules/commander/PreflightCheck.cpp]: leaking file descriptor - and so on. I'm fixing this right now.

**Comment by LorenzMeier** [comment]
@Zefz This is done now and passes bench testing. Could you verify this to make sure it still arms fine on a calibrated system?

**Comment by Zefz** [comment]
I noticed there is no feedback trying to arm while vehicle is in STANDBY_ERROR mode (no arming denied message or audio feedback).

After calibrating sensors from an invalid sensor state I get this message:
>>Critical: INVAL: ARMING_STATE_STANDBY_ERROR - ARMING_STATE_S
which means I cannot arm the system even though all sensors are now correctly calibrated.

A power cycle (reboot) fixes puts the system in the right state.

**Comment by LorenzMeier** [comment]
The missing feedback is obviously not good, but the lockdown into the standby error state is intended. Do you feel that's workable? The reason to do it this way is that I'm trying to enforce a power cycle to be sure we booted once nominally.

**Comment by Zefz** [comment]
Ah ok so this is an intended lockdown. This might not be apparent to the user, because QGC shows everything is green (even in STANDBY mode). I can see the safety switch rapidly blinking. Perhaps a single message to the user that the system needs a reboot will solve this?

**Comment by LorenzMeier** [comment]
Yes - if you feel like you could add a commit doing that and re-test it would be much appreciated. It seems to make sense to me to require a power cycle, just to avoid that we get into a weird state with a partially initialized system.

**Comment by Zefz** [comment]
I am getting the following mavlink message spammed to QGC while the preflight check fails:
> Critical: NOT ARMING: Sensors not operational.

Seems it is looping trying to enter a state and is denied because of the check failure?

**Comment by LorenzMeier** [comment]
@Zefz I've now added text feedback asking to power cycle for the corner case you mentioned. It would be great if you could re-test.

**Comment by LorenzMeier** [comment]
Applied to master, closing.

**Comment by Zefz** [comment]
Seems to be working good.

**Pull request closed without merging**